streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -16,128 +16,94 @@
+                    final TaskMovement taskMovement = new TaskMovement(task,  client);
+                    taskMovements.add(taskMovement);
+                }
+            }
+            clientsByTaskLoad.offer(client);
         }
 
-        int remainingAllowedWarmupReplicas = maxWarmupReplicas;
-        final List<TaskMovement> movements = new LinkedList<>();
-        for (final Map.Entry<UUID, List<TaskId>> sourceClientEntry : statefulActiveTaskAssignment.entrySet()) {
-            final UUID source = sourceClientEntry.getKey();
+        int remainingWarmupReplicas = maxWarmupReplicas;
+        for (final TaskMovement movement : taskMovements) {
+            final UUID leastLoadedClient = clientsByTaskLoad.poll(movement.task);
+            if (leastLoadedClient == null) {
+                throw new IllegalStateException("Tried to move task to caught-up client but none exist");
+            }
 
-            final Iterator<TaskId> sourceClientTasksIterator = sourceClientEntry.getValue().iterator();
-            while (sourceClientTasksIterator.hasNext()) {
-                final TaskId task = sourceClientTasksIterator.next();
-                final UUID destination = taskToDestinationClient.get(task);
-                if (destination == null) {
-                    log.error("Task {} is assigned to client {} in initial assignment but has no owner in the final " +
-                                  "balanced assignment.", task, source);
-                    throw new IllegalStateException("Found task in initial assignment that was not assigned in the final.");
-                } else if (!source.equals(destination)) {
-                    if (destinationClientIsCaughtUp(task, destination, tasksToCaughtUpClients)) {
-                        sourceClientTasksIterator.remove();
-                        statefulActiveTaskAssignment.get(destination).add(task);
-                    } else {
-                        if (clientStates.get(destination).prevStandbyTasks().contains(task)
-                                && tasksToRemainingStandbys.get(task) > 0
-                        ) {
-                            decrementRemainingStandbys(task, tasksToRemainingStandbys);
-                        } else {
-                            --remainingAllowedWarmupReplicas;
-                        }
+            final ClientState sourceClientState = clientStates.get(leastLoadedClient);
+            sourceClientState.assignActive(movement.task);
 
-                        movements.add(new TaskMovement(task, source, destination));
-                        if (remainingAllowedWarmupReplicas == 0) {
-                            return movements;
-                        }
-                    }
-                }
+            final ClientState destinationClientState = clientStates.get(movement.destination);
+            if (destinationClientState.prevStandbyTasks().contains(movement.task) && tasksToRemainingStandbys.get(movement.task) > 0) {
```

Comment:
Should it be:

```suggestion
            if (destinationClientState.previousAssignedTasks().contains(movement.task) && tasksToRemainingStandbys.get(movement.task) > 0) {
```
It doesn't seem like we care if the task was previously a standby or active, just that it wasn't caught up, and we have an available standby replica to give it, so we don't have to count it as a warmup.

Reply:
Q: Why do we even care at all whether the task was running on the client? What if we just assign a real stand-by task if we have a spare one?

Reply:
I think we're trying to strike a balance of not harming HA while moving tasks around.

So if we have a task T and three nodes A, B[T=active], C[T=standby], and we want to move T to A, then we'll do it by assigning a warmup to A and letting B continue to be active and C continue to be a standby until A catches up.